core/reactor_common.c

```diff
@@ -380,25 +372,25 @@ static token_freed _lf_free_token(lf_token_t* token) {
                 token->destructor(token->value);
             }
         }
-        token->value = NULL;
         result = VALUE_FREED;
+#endif
+        token->value = NULL;
     }
     // Tokens that are created at the start of execution and associated with
     // output ports or actions are pointed to by those actions and output
     // ports and should not be freed. They are expected to be reused instead.
-    if (token->ok_to_free) {
+    if (token->ok_to_free == token_and_value || token->ok_to_free == token_only) {
         // Need to free the lf_token_t struct also.
-        if (_lf_token_recycling_bin_size < _LF_TOKEN_RECYCLING_BIN_SIZE_LIMIT) {
+        if (hashset_num_items(_lf_token_recycling_bin) < _LF_TOKEN_RECYCLING_BIN_SIZE_LIMIT) {
             // Recycle instead of freeing.
-            token->next_free = _lf_token_recycling_bin;
-            _lf_token_recycling_bin = token;
-            _lf_token_recycling_bin_size++;
+            hashset_add(_lf_token_recycling_bin, token);
```

Comment:
based on the return value, we could put a comment if a duplicate entry is attempted to be added to hashset

Reply:
Good idea! This should result in a warning because a token should never be put on the recycling bin if it is already there.